epub33/core/index.html

```diff
@@ -436,14 +453,23 @@ <h3>Terminology</h3>
 						<dfn id="dfn-path-name" data-lt="Path Names">Path Name</dfn>
 					</dt>
 					<dd>
-						<p>For a given directory within the <a href="#sec-container-abstract">OCF Abstract
-							Container</a>, the string holding all directory <a>File Name</a> in the full path
-							concatenated together with a <code>/</code> (<code>U+002F</code>) character separating the
-							directory File Names.</p>
-						<p>For a given file within the OCF Abstract Container, the Path Name is the string holding all
-							directory File Names concatenated together with a <code>/</code> character separating the
-							directory File Names, followed by a <code>/</code> character and then the File Name of the
-							file.</p>
+						<p>For a given <var>file</var> within the <a href="#sec-container-abstract">OCF Abstract
+							Container</a> the Path Name can be obtained by performing the following steps (expressed using the terminology of [[infra]]):</p>
+
+						<ol>
+							<li>Let <var>path</var> be an empty <a data-cite="infra#list">list</a>.</li>
+							<li><a data-cite="infra#list-prepend">Prepend</a> the <a>File Name</a> of <var>file</var> to <var>path</var>.</li>
+							<li>Let <var>parent</var> be the parent directory of <var>file</var>.</li>
```

Comment:
I think this breaks if `file` is the root directory. In that case `parent` is undefined (or maybe null) and is also not the root directory, so the while condition doesn't terminate. Maybe move this line into the while body:

```
While file is not the Root Directory:
  let file be the parent of file
  prepend the File Name of file to path
  set file to the parent of file
```

Reply:
Good point. Alternatively if we don't want to mutate _file_, that becomes:

  1. Let _path_ be an empty list
  2. Let _parent_ be _file_
  2. While _parent_ is not the Root Directory:
    1. Prepend the File Name of _parent_ to _path_
    2. Set the parent directory of _parent_ to _parent_
  3. Return the concatenation of _path_ using the SOLIDUS (U+002F) character

Reply:
I am fine with that, but don't like `parent` as the name (parent of parent to parent is marginally confusing). How about item, part, fragment, or component?

Reply:
Yes, this was the most confusing piece to me on my first read through of the diff. I can't parse step 3. Doesn't there need to be a splitting of the file name from parent in 3.1, otherwise aren't you just setting parent to itself in step 3.2?

But the nitpicky editor point I wanted to make is that this really doesn't belong in a terminology definition. The definition should say what the term is, not define a parsing algorithm, especially if this is really a normative requirement to use somewhere. If there isn't a good place in the spec to put it, I'd consider splitting it out to an appendix.

Reply:
> Yes, this was the most confusing piece to me on my first read through of the diff. I can't parse step 3. Doesn't there need to be a splitting of the file name from parent in 3.1, otherwise aren't you just setting parent to itself in step 3.2?

I'm not sure I see what you mean. Step 3.2 is essentially

```
parent = parentOf(parent)
```

Of course, this assumes "the parent directory of" is a known operation (we do not formally specify what is a "file system").

That said, no problem changing the name of the variable to something else than _parent_ and easier to read, like @bduga suggests. I'm not a fan of `item`, `part`, `fragment` or `component` however, as the variable holds a file or directory (not a string, nor a path). Maybe `fileOrDir` ? or `f` ?

> 
> But the nitpicky editor point I wanted to make is that this really doesn't belong in a terminology definition. The definition should say what the term is, not define a parsing algorithm, especially if this is really a normative requirement to use somewhere. If there isn't a good place in the spec to put it, I'd consider splitting it out to an appendix.

Right, I agree 😊 (see [my earlier comment](https://github.com/w3c/epub-specs/pull/1898#discussion_r746609216))

Reply:
> Of course, this assumes "the parent directory of" is a known operation

Exactly. I couldn't get that from the description. It's just words that seem to say that the parent directory of parent is itself. There's no operation there.

But either way works, I assume. You could have a single function in 3.1 that returns the directory name and remaining path or you could get the directory name in one step and then the remaining path in the next.

> Right, I agree 😊 (see [my earlier comment](https://github.com/w3c/epub-specs/pull/1898#discussion_r746609216))

Ah, missed that one!

Reply:
Could be the ordering of words that's confusing me here: "Set _parent_ to the parent directory of _parent_" would be clearer.

Reply:
Still too many parents in one line for me, but also parent is no longer correct since it can be `file`, which can't be a parent. I am fine with item or fileOrDir, or even something simpler like current, curr, c, i, f, etc.

Reply:
> You could have a single function in 3.1 that returns the directory name and remaining path or you could get the directory name in one step and then the remaining path in the next.

I'm not sure I see what you mean 😊.
Currently the algorithm uses the "File Name of" operation (takes a file, returns a string), which is defined in the terminology, and "parent directory of" operation (takes a file, returns a file), which is assumed to be well-known for a file system. 
Are you suggesting replacing that with a single function that takes a file and returns a tuple (string, file)? 

> "Set parent to the parent directory of parent" would be clearer

Ouch 🤦 , you're right that's the correct way to write it . Sorry for all the confusion (in my mind I thought "assign value to variable", when it's "set variable to value" 😅).

Reply:
> Still too many parents in one line for me, but also parent is no longer correct since it can be `file`, which can't be a parent. I am fine with item or fileOrDir, or even something simpler like current, curr, c, i, f, etc.

I like `current` 😊

Reply:
> Currently the algorithm uses the "File Name of" operation (takes a file, returns a string), which is defined in the terminology,

It doesn't, though. It links to the "file name" definition, not the "path name" definition with the algorithm. Fix that and it's now becoming clear to me... 😄

Reply:
> It doesn't, though. It links to the "file name" definition, not the "path name" definition with the algorithm. Fix that and it's now becoming clear to me... 😄

Now _I_'m confused 😁. It does mean to link to the "file name" definition! Basically, the algorithm loops from the input file to all its parent directories and make a list of **file names** (each file-or-directory name is one path component), which it then concatenates to build a path.

Reply:
> Now _I_'m confused 😁.

Ya, sorry, I was trying to make Amelia dinner and respond. Didn't work out well... 😛 

> It does mean to link to the "file name" definition

The file name definition is just a prose definition that says: "The name of any type of file within an OCF Abstract Container, whether a directory or a file within a directory."

Is the assumption that the algorithm doesn't need to define what it actually means to split `parent` to obtain the file name, or isn't something missing here? (I didn't mean to link to "path name", but that it doesn't have a similar algorithm.)

Sort of similar to how "get the parent directory" isn't really defined, but obviously if you split off the last segment you'll get the parent. How abstract/detailed is the algorithm supposed to be, in other words?

In that light, to try and explain my original question better, if you're splitting a path to obtain the current directory/file name, and detailing how this is done, and you also want what remains of the input string because it represents the new parent, do you really need to run the same operation twice?

Couldn't you write it along these lines (overlooking probably awful infra as I don't have the energy to reread that spec right now!):

Let `parsed` be a list
Set `parsed` to the result of running `parseFileDir` with `current` as an argument
Prepend parsed[0] to Path
Set current to parsed[1]

and then define parseFileDir (or whatever name makes sense) to split on the last instance of the path separator and return the two components as a list?

That's how I had this conceived in my head as I read it, at any rate.

Reply:
> Is the assumption that the algorithm doesn't need to define what it actually means to split `parent` to obtain the file name, or isn't something missing here? (I didn't mean to link to "path name", but that it doesn't have a similar algorithm.)

The assumption of the algorithm is that getting the File Name of a container file or directory is well-defined. If we think it is not, then we have to further detail the terminology entry of "File Name".

Now, I think the confusion comes from the nature of the variable _parent_, or rather, let's use _current_. It is a **file**, not a **path string**. There cannot be any "splitting" involved, since you don't want to split a file. If you want to split a string, what string are you splitting? There is no "path" or "path separator" that can be used already, since this is all what we're currently defining in the "Path Name" algorithm.
If we had a path string with known separators, we wouldn't need to define an algorithm to define that path string in the first place 😁.

In fact, I originally suggested to use a proper Infra algorithm to define "Path Name" as the original definition was kinda algorithmic: 
> For a given file within the OCF Abstract Container, the Path Name is the string holding all directory File Names concatenated together with a / character separating the directory File Names, followed by a / character and then the File Name of the file.

By contrast, the definition of "File Name" is not algorithmic but axiomatic. It assumes that within a file system, a file has a name. An implicit axiom (of the existence of a file system) is that a file has a parent directory.

In fact I can see two approaches:

1. purely axiomatic:
  * we assume a file/directory has a "name" (as a well-known concept)
  * we assume a file/directory has a "path" (as a well-known concept)
  * in the definition of "Path Name", we just have to specify that the path uses "/" as a separator
2. axiomatic + algorithmic
  * we assume a file/directory has a "name" (as a a well-known concept)
  * we assume a file/directory has a parent directory (as a well-known concept)
  * we define "Path Name" algorithmically using the two assumptions above

I understood the original "Path Name" definition as closer to (2) than (1), so I suggested to refine the algorithmic part.

That said, a purely axiomatic approach might work too.

Reply:
> It is a **file**, not a **path string**.

Ah, okay, that's not how I read it, obviously. I'd assumed this was trying to now define how to normalize any file path (e.g., in an `href` or from the zip path).

But if the abstract container is a virtual file system, can't you assume your axiomatic approach that only requires a slash as the separator? (i.e., if you can get the parent directories, then doesn't a path exist?)

I see what you mean now about the original sounding algorithmic, but at the same time it also seems very much like it's only standardizing the use of slashes as separators.

Reply:
> it also seems very much like it's only standardizing the use of slashes as separators.

Essentially, yes. It also ensures the path doesn't start with a slash.

> can't you assume your axiomatic approach that only requires a slash as the separator?

Maybe. But then we need to come up with a better definition than the old one, which is too vague IMO (algorithmic in some way, but not precise).

Reply:
> But then we need to come up with a better definition than the old one, which is too vague IMO (algorithmic in some way, but not precise).

Sure, I'm not advocating for a return to the old wording.

It's one of those bits you get the general gist of despite not being perfectly clear, so no one's really dug into it in 3.x.

Reply:
See https://github.com/w3c/epub-specs/pull/1898#issuecomment-966251807

Is it o.k. to mark this as resolved (and comment on the latest iteration in https://github.com/w3c/epub-specs/pull/1898#issuecomment-966251807)?

Reply:
> Is it o.k. to mark this as resolved

I'm having second thoughts and would like to explore the axiomatic approach (see comment above for what that means 😁).

I searched existing Web standards for file-related terminology, and found at least:
- [File and Directory Entries API](https://wicg.github.io/entries-api/), notably its [concepts section](https://wicg.github.io/entries-api/#concepts) that looks very similar to what we're defining.
- [Directory Upload](https://wicg.github.io/directory-upload/proposal.html) and its [Model section](https://wicg.github.io/directory-upload/proposal.html#directory-file-root-and-path) 
- [File System Access](https://wicg.github.io/file-system-access/) and its [Concepts](https://wicg.github.io/file-system-access/#concepts) section
- [File API](https://w3c.github.io/FileAPI/)

My intent with the algorithm was to clarify the original definition that was somehow algorithmic. But:
 * we have to rely on axioms at some point. The existence of a `Path Name` could be one of these. As @mattgarrish said, this has been left unnoticed for years, so there's low risk of misinterpretation. 
 * ideally, there would be a Web standard for a common "File System" model, that could be referenced by all the above APIs and by EPUB. This is WICG territory however.
 * maybe it's not worth having a clear algorithm, when the other definitions (file, file name, parent directory, etc) are not so well-defined.

So an alternative to the "Path Name" algorithm would be:
- possibly define a little more clearly the OCF file system model: container contains files and directories, they have a name, a path, etc.
- reword the definition of "Path Name", notably the algorithmic-sounding wording.

That could be a distinct PR (note: I don't have the time to work on this in the coming days 😅).

Reply:
We agreed (see https://github.com/w3c/epub-specs/pull/1898#issuecomment-966904662) to raise a separate issue on whether we use an algorithm or a declarative form. Apart from that, I believe this is now resolved.